Doc/library/asyncio-sync.rst

```diff
@@ -356,6 +357,135 @@ BoundedSemaphore
       :ref:`What's New in 3.10's Removed section <whatsnew310-removed>`
       for more information.
 
+
+Barrier
+=======
+
+   .. versionadded:: 3.10
+
+   A barrier object.  Not thread-safe.
+
+   This class, as a clone of :class:`threading.Barrier`, provides a simple synchronization
+   primitive for use by a fixed number of tasks that need to wait for each other.
+   Each of the task tries to pass the barrier by calling the :meth:`~Barrier.wait` method
+   and will block until all of the tasks have made their :meth:`~Barrier.wait` calls.
+   At this point, the tasks are released simultaneously.
+
+   The barrier can be reused any number of times for the same number of tasks.
+
+   .. _asyncio_example_barrier:
+
+   Example::
+
+       import asyncio
+       import os
+       import random
+
+       DELAYS = (1.0, 4.0, 2.0, 0.5)
+       NB = 10
+
+       def job_done():
+          msg = 'ACTION: All files removed'
+          print(f'{os.linesep}{msg:-^80s}{os.linesep}')
+
+       async def remove_file(barrier, filename):
+          print(f'Remove {filename!r} ...')
+
+          # remove filename (simulation)
+          await asyncio.sleep(random.choice(DELAYS))
+          print(f'\tFile {filename!r} is removed')
+
+          # wait for others files
+          return await barrier.wait()
+
+       async def remove_folder(barrier, folder):
+          print(f'Waiting for remove all files from folder {folder!r} ...')
+
+          # wait for all removed files
+          await barrier.wait()
+
+          # remove folder (simulation)
+          await asyncio.sleep(1.0)
+          print(f'Folder {folder!r} is now removed ...')
+
+          return -1
+
+       async def main(folder, files, nb_files):
+          # Create a Barrier object with parties equal to nb_files+1
+          # corresponding to the number of all removing items,
+          # and an action as a print message.
+          barrier = asyncio.Barrier(nb_files+1, action=job_done)
+
+          # Add remove_file task for each filename.
+          tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]
+
+          # Add remove_folder task for the folder.
+          tasks.append(asyncio.create_task(remove_folder(barrier, folder)))
+
+          # Wait until folder+filenames are removed.
+          await asyncio.gather(*tasks)
+
+       asyncio.run(main("Datas", [f'file{i+1:02d}.txt' for i in range(NB)], NB))
+
+.. class:: Barrier(parties, action=None)
+
+   Create a barrier object for *parties* number of tasks.  A callable *action*, when
+   provided, is called once just before the release phasis (draining state).
```

Comment:
I'm wondering if the design of this primitive should deviate from the threading version. In this case, I think we should rely on `async with` instead of passing a callback (which are always a PITA to deal with in async/await context) :

```python
async with asyncio.Barrier(10) as b:
    tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]
    tasks.append(asyncio.create_task(remove_folder(barrier, folder)))
    await asyncio.gather(*tasks)

msg = 'ACTION: All files removed'
print(f'{os.linesep}{msg:-^80s}{os.linesep}')
```

Reply:
Thank you @1st1 for your response. 

Notice: when specified, the callback runs **once** from the last task calling the  `wait` method and just before draining the barrier. If the callback crashes, the barrier state goes to broken.

Are you suggesting to remove that callback parameter or to replace with a coro and run an `awaiting` of it ? I agree to respect the asyncio context, but in that case I don't understand a PITA with this callback. May I miss something ?

Moreother, `__aenter__` and `__aexit__`  methods  are missing in the Barrier class. To add with dedicated tests.